eva-server/src/main/java/uk/ac/ebi/eva/server/ws/ga4gh/GA4GHBeaconWSServerV2.java

```diff
@@ -0,0 +1,325 @@
+                            null));
+                } else {
+                    datasetAllelResponses.add(new DatasetAlleleResponse(variantSourceMongo.getStudyId(),
+                            false,
+                            null,
+                            new Float(0),
+                            variantSourceMongo.getStats() == null ? null :
+                                    new Long(variantSourceMongo.getStats().getVariantsCount()),
+                            null,
+                            variantSourceMongo.getStats() == null ? null :
+                                    new Long(variantSourceMongo.getStats().getSamplesCount()),
+                            "noteString",
+                            "externalUrl",
+                            null));
+
+                }
+            });
+
+        } else if (request.getIncludeDatasetResponses().equalsIgnoreCase("HIT")) {
+            allStudies.forEach((studyId, variantSourceMongo) -> {
+                if (studiesPresent != null && studiesPresent.contains(studyId)) {
+                    datasetAllelResponses.add(new DatasetAlleleResponse(variantSourceMongo.getStudyId(),
+                            true,
+                            null,
+                            studyIdToFrequencyMapper.get(variantSourceMongo.getStudyId())==null?new Float(0):
+                                    new Float(studyIdToFrequencyMapper.get(variantSourceMongo.getStudyId())),
+                            variantSourceMongo.getStats() == null ? null :
+                                    new Long(variantSourceMongo.getStats().getVariantsCount()),
+                            null,
+                            variantSourceMongo.getStats() == null ? null :
+                                    new Long(variantSourceMongo.getStats().getSamplesCount()),
+                            "noteString",
+                            "externalUrl",
+                            null));
+                }
+            });
+        } else if (request.getIncludeDatasetResponses().equalsIgnoreCase("MISS")) {
+            allStudies.forEach((studyId, variantSourceMongo) -> {
+                if (studiesPresent != null && !studiesPresent.contains(studyId)) {
+                    datasetAllelResponses.add(new DatasetAlleleResponse(variantSourceMongo.getStudyId(),
+                            false,
+                            null,
+                            new Float(0),
+                            variantSourceMongo.getStats() == null ? null :
+                                    new Long(variantSourceMongo.getStats().getVariantsCount()),
+                            null,
+                            variantSourceMongo.getStats() == null ? null :
+                                    new Long(variantSourceMongo.getStats().getSamplesCount()),
+                            "noteString",
+                            "externalUrl",
+                            null));
+                }
+            });
```

Comment:
after the prototyping stage, I would recommend extracting these `new DatasetAlleleResponse`. Also, I think it would be less redundant if you structure this part (70 lines) as:
```
allStudies.forEach(study -> {
    if  studiesPresent.contains(study) {
        if ALL || HIT {
            responses.add(buildResponse(true, variantSourceMongo)
        }
    } else {
         if ALL || MISS {
            responses.add(buildResponse(false, variantSourceMongo)
        }
    }
})

Reply:
That's great!!
I will change it to this format

Reply:
Fixed